spi/src/main/java/org/opensearch/ml/common/spi/tools/Tool.java

```diff
@@ -9,7 +9,7 @@
 import java.util.Map;
 
 /**
- * General tool interface.
+ * General tool interface. Tool implementation needs to be thread safe.
```

Comment:
> Tool implementation needs to be thread safe.

This is a high requirement for tool developer. Before this refactoring , we don't require this as we create new Tool for each request. I would suggest keep it simpler for tool developer.

Reply:
Agreed that tool developers needs to be careful with maintaining global states  but Singleton design pattern is a standard design pattern and has the benefit of using less memory overhead. Tool implementation needs to be stateless and should only perform the operation of calling other plugins/opensearch to fetch required information. If tools could be stateless do you see any concerns of maintaining a singleton tool?

Reply:
Seems you missed this comment https://github.com/opensearch-project/ml-commons/pull/1636#discussion_r1406940721